src/api/Nhs.Appointments.Core/AllocationStateService.cs

```diff
@@ -0,0 +1,108 @@
+            {
+                targetSlot.Capacity--;
+                
+                //TODO this still is only needed for when running a recalculation
+                //waste of memory etc for when no recalculations required
+                allocationState.SupportedBookingReferences.Add(booking.Reference);
+            }
+        }
+
+        allocationState.AvailableSlots = slots.Where(s => s.Capacity > 0).ToList();
+
+        return allocationState;
+    }
+
+    /// <summary>
+    ///     'Greedy' solution for assigning bookings to slots for multiple services in a deterministic way
+    /// </summary>
+    /// <param name="slots"></param>
+    /// <param name="booking"></param>
+    /// <returns></returns>
+    private static SessionInstance ChooseHighestPrioritySlot(IEnumerable<SessionInstance> slots, Booking booking) =>
+        slots.Where(sl => sl.Capacity > 0
+                          && sl.From == booking.From
+                          && (int)sl.Duration.TotalMinutes == booking.Duration
+                          && sl.Services.Contains(booking.Service))
+            .OrderBy(slot => slot.Services.Length)
+            .ThenBy(slot => string.Join(string.Empty, slot.Services.Order()))
+            .FirstOrDefault();
+
+    private async Task<IEnumerable<Booking>> GetBookings(string site, DateTime from, DateTime to) =>
+        await bookingQueryService.GetOrderedLiveBookings(site, from, to);
+
+    private async Task<IEnumerable<SessionInstance>> GetSlots(string site, DateTime from, DateTime to) =>
+        await availabilityQueryService.GetSlots(site, DateOnly.FromDateTime(from), DateOnly.FromDateTime(to));
+}
+
+public static class RecalculationExtensions
+{
+    public static void AddNewlySupportedBookings(this List<BookingAvailabilityUpdate> recalculations, AllocationState state, List<Booking> bookings)
+    {
+        recalculations.AddRange(bookings
+            .Where(booking => state.SupportedBookingReferences.Contains(booking.Reference) &&
+                              booking.AvailabilityStatus is not AvailabilityStatus.Supported)
+            .Select(booking => new BookingAvailabilityUpdate(booking, AvailabilityUpdateAction.SetToSupported)));
+    }
+
+    public static void AddNoLongerSupportedBookings(this List<BookingAvailabilityUpdate> recalculations, AllocationState state, List<Booking> bookings)
+    {
+        recalculations.AddRange(bookings
+            .Where(booking => !state.SupportedBookingReferences.Contains(booking.Reference) &&
+                              booking.AvailabilityStatus is AvailabilityStatus.Supported &&
+                              booking.Status is not AppointmentStatus.Provisional)
+            .Select(booking => new BookingAvailabilityUpdate(booking, AvailabilityUpdateAction.SetToOrphaned)));
+    }
+
+    public static void DeleteRemainingUnsupportedProvisionalBookings(this List<BookingAvailabilityUpdate> recalculations, AllocationState state, List<Booking> bookings)
```

Comment:
This method is named differently from the others, which are `AddFoo...`. The name difference made me assume this method was actually doing the delete